git101/index.md

```diff
@@ -0,0 +1,301 @@
+- Any change can be made to the repository copy
+- Changes can be synchronized between copies
+- Repository synchronization requires both sides to agree on the __exact__
+series of events in their history
+  - There are __NO__ exceptions to this rule
+- Branch-merge strategy allows private changes that can later be merged into
+a common history in an orderly manner
+
+---
+# What’s the alternative?
+- Centralized systems like Subversion are also available
+- Repositories are cached locally, but are not authoritative copies
+- Committing requires a connection to the central repository
+- Much less flexibility, but much less chance of disagreement
+
+---
+# Git ecosystem survey
+- StackOverflow conducts an annual survey of developer trends
+- Until 2018 they asked about version control preferences
+- Now they just ask what services you use with git
+
+---
+![height:15cm](vc-tools.png)
+<https://insights.stackoverflow.com/survey/2018>
+
+---
+![height:15cm](commit-freq.png)
+<https://insights.stackoverflow.com/survey/2018>
+
+---
+![height:15cm](collab-tools.png)
+<https://insights.stackoverflow.com/survey/2020>
+
+---
+# So what about GitHub?
+- A social network for git users
+- A place to keep a backup repository copy
+- A means to exchange changes between people
+- A database of issues
+- A database of change requests, comments, and approvals
+- An API for running events on repository changes
+
+---
+# So what about GitHub?
+- Remember that git is a tool, and GitHub is a provider of git services
+- Other services:
+  - Azure Repos, Google Cloud Source, AWS CodeCommit
+  - Bitbucket
+  - GitLab
+- Other git implementations:
+  - Eclipse EGit
+  - GitHub Desktop
+  - Git Bash (ok, so it's just a port)
+- A [student pack](https://education.github.com/pack) is available
+
+---
+# What is git?
+
+- Git stores a series of changes to files and attributes of that change
+  - File diff, author, timestamp, previous change hash
+- Change and attributes are hashed together to uniquely identify a commit
+- Branches allow different authors to pursue different paths
+  - Like an actual tree, branches can go their own way, but the trunk must be shared
+- Never delete a commit: use a revert to undo, but that change is then logged
+- Some leeway if your changes haven't been pushed and become visible to others
+
+---
+# A note on branching and language
+
+- Git-based projects have a primary branch, traditionally known as `master`
+- You can develop new code on different branches, with the goal of integrating
+that code back into the primary branch
+- Terminology and language matter, and git's use of the term `master` is
+considered insensitive and has been deprecated
+- Consider naming your primary branch `main` or `develop` on new projects
+
+---
+# Help along the way
+
+- Git uses man pages extensively to document its options and features
+- One man page would be too much for all of git, so pages are hyphenated with
```

Comment:
Each command also has a `--help` flag which usually opens up that particular commands `man` page.

Reply:
No GNU options allowed here